### PR TITLE
feat(security): implement stream key authentication

### PR DESCRIPTION
**Pull request by onyedikachi-david**
/claim #41
Fixes: #41

### CODE REVIEW

File: lib/algora/accounts.ex [hunk 1]
```diff
@@ -5,6 +5,8 @@ defmodule Algora.Accounts do
   alias Algora.{Repo, Restream}
   alias Algora.Accounts.{User, Identity, Destination, Entity}
 
+  @stream_key_expiration_days 30
```
Comment (zcesur) [inline lib/algora/accounts.ex:8]: Why are we expiring stream keys? They should be valid forever unless the user resets

File: lib/algora/pipeline/message_validator.ex [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  defp validate_stream_key(user, %{stream_key: key}) when is_binary(key) do
+    case Accounts.validate_stream_key(user, key) do
+      {:ok, user} -> {:ok, user}
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  defp validate_stream_key(_user, _message), do: {:error, :missing_stream_key}
+
+  def authenticate_user(%{username: username, password: password}) do
```
Comment (zcesur) [inline lib/algora/pipeline/message_validator.ex:22]: Are we sure the message struct contains these fields?

**Review by zcesur — CHANGES_REQUESTED** [review]
Thanks for the PR. Once it's done can you share a short demo of the feature?

**Comment by onyedikachi-david** [comment]
> Thanks for the PR. Once it's done can you share a short demo of the feature?

Hi, I have been trying to run locally with these setups. I tried with these docker-compose modifications and seeds since I need to be authenticated to see the settings, but I can't access the channel/settings endpoint. Please, what is your setup for local development?


<details>
<summary><strong>docker-compose.yml</strong></summary>
```yml
services:
  db:
    image: postgres:13
    environment:
      POSTGRES_USER: dev_user
      POSTGRES_PASSWORD: dev_password
      POSTGRES_DB: dev_db
    ports:
      - "5432:5432"
    volumes:
      - postgres_data:/var/lib/postgresql/data

  app:
    build:
      context: .
      dockerfile: Dockerfile
    environment:
      MIX_ENV: dev
      DATABASE_URL: "postgres://dev_user:dev_password@db:5432/dev_db"
      SEED_DB: "true"  # Set this to "true" to run seeds
    ports:
      - "4000:4000"
    depends_on:
      - db
    command: ["phx.server"]
    volumes:
      - .:/app
      - deps:/app/deps
      - _build:/app/_build

volumes:
  postgres_data:
  deps:
  _build:
```
</details>
<details>
<summary><strong>Dockerfile</strong></summary>

```yml
# Base images
ARG BUILDER_IMAGE="hexpm/elixir:1.15.7-erlang-26.2-debian-bookworm-20231009-slim"
ARG RUNNER_IMAGE="hexpm/elixir:1.15.7-erlang-26.2-debian-bookworm-20231009-slim"

FROM ${BUILDER_IMAGE} as builder

# Install build dependencies
RUN apt-get update -y && \
    apt-get install -y build-essential git curl postgresql-client && \
    apt-get clean && \
    rm -f /var/lib/apt/lists/*_*

# Prepare build dir
WORKDIR /app

# Install hex + rebar
RUN mix local.hex --force && \
    mix local.rebar --force

# Set build ENV to development
ENV MIX_ENV="dev"

# Install mix dependencies
COPY mix.exs mix.lock ./
RUN mix deps.get --only $MIX_ENV
RUN mkdir config

# Copy compile-time config files
COPY config/config.exs config/${MIX_ENV}.exs config/
RUN mix deps.compile

COPY priv priv

# Compile the release
COPY lib lib

COPY assets assets
RUN mix assets.deploy
RUN mix compile

# Copy runtime config
COPY config/runtime.exs config/
COPY rel rel

# Don't run mix release for dev
# RUN mix release

FROM ${RUNNER_IMAGE}

# Install runtime dependencies
RUN apt-get update -y && \
    apt-get install -y libstdc++6 openssl libncurses5 locales ffmpeg imagemagick postgresql-client inotify-tools && \
    apt-get clean && \
    rm -f /var/lib/apt/lists/*_*

WORKDIR /app

# Copy the built artifact from the builder stage
COPY --from=builder /app /app
COPY --from=builder /root/.mix /root/.mix

# Install hex and rebar
RUN mix local.hex --force && \
    mix local.rebar --force

# Add this near the end of the file
COPY entrypoint.sh /app/
RUN chmod +x /app/entrypoint.sh

# Change the ENTRYPOINT to use shell form
ENTRYPOINT ["/bin/bash", "/app/entrypoint.sh"]
CMD ["phx.server"]
```
</details>
<details>
<summary><strong>seeds.exs</strong></summary>

```elixir
alias Algora.Library.Video
alias Algora.Repo
# Script for populating the database. You can run it as:
#
#     mix run priv/repo/seeds.exs
#
# Inside the script, you can read and write to any of your
# repositories directly:
#
#     Algora.Repo.insert!(%Algora.SomeSchema{})

# Uncomment and modify as needed

# Then modify your insert statement:

# At the top of your seeds.exs file:
# alias Algora.Library.Video
# alias Algora.Repo

for i <- 1..200 do
  type = if rem(i, 2) == 0, do: :vod, else: :livestream
  format = if type == :vod, do: :mp4, else: :hls

  video_attrs = %{
    title: "Sample #{if type == :vod, do: "Video", else: "Livestream"} #{i}",
    description: "This is a sample #{if type == :vod, do: "video", else: "livestream"} description.",
    type: type,
    format: format,
    is_live: type == :livestream,
    visibility: :public,
    duration: :rand.uniform(3600)  # Random duration up to 1 hour (3600 seconds)
  }

  changeset =
    %Video{}
    |> Video.changeset(video_attrs)
    |> Video.put_video_url(format)

  # Manually set fields that might not be included in the changeset
  changeset = Ecto.Changeset.force_change(changeset, :type, type)
  changeset = Ecto.Changeset.force_change(changeset, :format, format)

  case Repo.insert(changeset) do
    {:ok, video} ->
      IO.puts("Created video: #{video.id} - #{video.title}")
    {:error, changeset} ->
      IO.puts("Failed to create video:")
      IO.inspect(changeset.errors)
  end
end
```

</details>
<details>
<summary><strong>Log output</strong></summary>

```cmd
app-1  | Created video: 392 - Sample Video 188
app-1  | Created video: 393 - Sample Livestream 189
app-1  | Created video: 394 - Sample Video 190
app-1  | Created video: 395 - Sample Livestream 191
app-1  | Created video: 396 - Sample Video 192
app-1  | Created video: 397 - Sample Livestream 193
app-1  | Created video: 398 - Sample Video 194
app-1  | Created video: 399 - Sample Livestream 195
app-1  | Created video: 400 - Sample Video 196
app-1  | Created video: 401 - Sample Livestream 197
app-1  | Created video: 402 - Sample Video 198
app-1  | Created video: 403 - Sample Livestream 199
app-1  | Created video: 404 - Sample Video 200
app-1  | Starting Phoenix app...
app-1  | WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
app-1  | I0000 00:00:1723949312.094452     146 tfrt_cpu_pjrt_client.cc:349] TfrtCpuClient created.
app-1  | [info] Detected running on primary. No local replication to track.
app-1  | [info] Running AlgoraWeb.Endpoint with cowboy 2.10.0 at 0.0.0.0:4000 (http)
app-1  | [info] Access AlgoraWeb.Endpoint at http://localhost:4000
app-1  | [info] Running AlgoraWeb.Embed.Endpoint with cowboy 2.10.0 at 0.0.0.0:4001 (http)
app-1  | [info] Access AlgoraWeb.Embed.Endpoint at http://localhost:4001
app-1  | [watch] build finished, watching for changes...
app-1  | 
app-1  | Rebuilding...
app-1  | 
app-1  | Done in 15731ms.
```
</details>

**Comment by zcesur** [comment]
Hey David, the only way to authenticate right now is to [create a GitHub OAuth app](https://github.com/settings/applications/new) (takes less than a minute) with

> Homepage URL: http://localhost:4000
> Authorization callback URL: http://localhost:4000/oauth/callbacks/github

and then set `GITHUB_CLIENT_ID` and `GITHUB_CLIENT_SECRET` with your app's values inside your .env file

**Comment by onyedikachi-david** [comment]
Should I create a pull request for the dev setup i used for run locally? It was really difficult to come up with a local dev setup. Skill issue though :) @zcesur. I also will be closing this pull request; I'm creating another.

**Comment by zcesur** [comment]
Sounds good. Was there anything in particular that gave you trouble? Are you running a docker based setup now or a bare mix based one? Feel free to submit your setup, would love to make it easier to contribute!

**Comment by onyedikachi-david** [comment]
> Sounds good. Was there anything in particular that gave you trouble? Are you running a docker based setup now or a bare mix based one? Feel free to submit your setup, would love to make it easier to contribute!

A docker based set-up. I'm thinking of making it a devcontainer...

**Comment by onyedikachi-david** [comment]
I can't remember all the errors I had then.

**Pull request closed without merging**